File: 程序员面试金典/2.4分割链表/partition.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

// 创建一个新链表，将小于x的节点加入头部，大于等于x的节点加入尾部
class Solution {
    public ListNode partition(ListNode head, int x) {
        if (head == null)
            return head;
        ListNode newHead = head, tail = head, p = head;
        while (p != null) {
            ListNode next = p.next;
            if (p.val < x) {
                p.next = newHead;
                newHead = p;
            }
            else {
                tail.next = p;
                tail = p;
            }
            p = next;
        }
        tail.next = null;
        return newHead;
    }
}

File: 程序员面试金典/2.4/partition.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# 将小于x的节点移动到前面
class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        new_head = ListNode(0)
        new_head.next = head
        p = new_head
        while p.next is not None:
            if p.next.val < x and p is not new_head:
                node = p.next
                p.next = node.next
                node.next = new_head.next
                new_head.next = node
            else:
                p = p.next
        return new_head.next
